clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -337,24 +404,34 @@ func (p *pingChecker) run() {
 				time.Sleep(pingBackoff)
 				continue
 			}
+			msg := ""
 			switch p.svc.Spec.Type {
 			case corev1.ServiceTypeClusterIP:
-				cmd := fmt.Sprintf("curl %s:%d", p.svc.Spec.ClusterIP, p.svc.Spec.Ports[0].Port)
-				_, err = execservice.RunCommand(pod, cmd)
+				// curl parameter is https://www.mit.edu/afs.new/sipb/user/ssen/src/curl-7.11.1/docs/curl.html
+				// we use 3 as the value of -m, instead of the default timeout value 120s, to make the service creation time more precise
+				cmd := fmt.Sprintf("curl -m 3 -s -S %s:%d", p.svc.Spec.ClusterIP, p.svc.Spec.Ports[0].Port)
+				msg, err = execservice.RunCommand(pod, cmd)
 			case corev1.ServiceTypeNodePort:
-				cmd := fmt.Sprintf("curl %s:%d", pod.Status.HostIP, p.svc.Spec.Ports[0].NodePort)
-				_, err = execservice.RunCommand(pod, cmd)
+				cmd := fmt.Sprintf("curl -m 3 -s -S %s:%d", pod.Status.HostIP, p.svc.Spec.Ports[0].NodePort)
+				msg, err = execservice.RunCommand(pod, cmd)
 			case corev1.ServiceTypeLoadBalancer:
-				cmd := fmt.Sprintf("curl %s:%d", p.svc.Status.LoadBalancer.Ingress[0].IP, p.svc.Spec.Ports[0].Port)
-				_, err = execservice.RunCommand(pod, cmd)
+				cmd := fmt.Sprintf("curl -m 3 -s -S%s:%d", p.svc.Status.LoadBalancer.Ingress[0].IP, p.svc.Spec.Ports[0].Port)
+				msg, err = execservice.RunCommand(pod, cmd)
 			}
 			if err != nil {
 				success = 0
 				time.Sleep(pingBackoff)
 				continue
 			}
+			if p.regexString != "" {
```

Comment:
This part is supercryptic. I do not know what should we expect from curl and what are we looking for in this regexp.

I *guess* that we want to get additional IPs for this service, but why do we use this format? What's the assumption for curl output?

Reply:
I agree with you. This block code here is not straightforward. One of the main reason is that the regexpattern is replying on the output of the container_image when gets curled.

For example, I am using gcr.io/eminent-nation-87317/webserver:v1, which output is similar as below in the k8s cluster. This image could echo its serving pods's ip in the output.  I am using this image as test pod to check whether multiple different pod-ip is curled through the clusterip service(as well as nodeport/lb service). regexString is not mandatory, tester could leverage it according to their test purpose and need to set appropriate regexstring pattern align with their test pod image's output. 

Yes it is tricky when you see code such as "len(ip) >= 2", it is also because of the regexString for my test pod is "Server address: ([0-9.]+):80". 2 could guarantee that we find the pod ip in the curl's output. I also would like the code to be generic and instrinct,  but I do not get perfect solution. If possible, would you please share more throughts/suggestions here. Really appreciate. 

```
root@harbor:~[62460]# kubectl get pods -o wide
NAME                        READY   STATUS      RESTARTS   AGE    IP              NODE                               NOMINATED NODE   READINESS GATES
curl                        0/1     Completed   0          5d3h   100.96.34.234   cls-1c-100w-md-0-5b5db699f-x8q44   <none>           <none>
test-dep-6c9c74b86d-2qdg4   1/1     Running     0          5d4h   100.96.55.254   cls-1c-100w-md-0-5b5db699f-js78x   <none>           <none>
test-dep-6c9c74b86d-d4jvq   1/1     Running     0          5d4h   100.96.48.251   cls-1c-100w-md-0-5b5db699f-t9s8l   <none>           <none>
test-dep-6c9c74b86d-t76kf   1/1     Running     0          5d4h   100.96.76.248   cls-1c-100w-md-0-5b5db699f-4tkhj   <none>           <none>

root@harbor:~[62456]# kubectl get svc test-svc
NAME       TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
test-svc   ClusterIP   100.66.21.80   <none>        80/TCP    5d4h

root@harbor:~[62456]# kubectl run curl2 --image curlimages/curl --restart=Never  -- sleep 36000
pod/curl2 created
root@harbor:~[62457]# kubectl exec -it curl2 -- curl 100.66.21.80
Server address: 100.96.76.248:80
Server name: test-dep-6c9c74b86d-t76kf
URI: /
root@harbor:~[62458]# kubectl exec -it curl2 -- curl 100.66.21.80
Server address: 100.96.55.254:80
Server name: test-dep-6c9c74b86d-2qdg4
URI: /
root@harbor:~[62458]# kubectl exec -it curl2 -- curl 100.66.21.80
Server address: 100.96.55.254:80
Server name: test-dep-6c9c74b86d-2qdg4
URI: /
```